FRONTEND.md

```diff
@@ -0,0 +1,533 @@
+argument of the current program.
+
+### DBIdx
+
+`DBIdx(val)` are input type references. They are allowed in the shape values of `DShapedArray`
+objects, found in the *in_type* signature of Python/Jaxpr programs. The integer value of a
+reference are interpreted as an index in the current `in_type` signature tuple.
+
+Input type indices are:
+
+- Produced when calculating the input type of a nested program [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L2321)
+- Consumed when creating argument tracers of a nested program [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L2437)
+
+### InDBIdx/OutDBIdx
+
+`InDBIdx(val)/OutDBIdx(val)` are output type references. references allowed in the shape values of `DShapedArray`
+objects found in the *out_type* signatures of Python/Jaxpr programs.
+
+- *InDBIdx(val)* refers to the position in the `in_type` signature tuple of the Jaxpr/Python
+  program.
+- *OutDBIdx(val)* refers to the position in the `out_type` signature tuple of the Jaxpr/Python
+  program.
+
+Output type indices are:
+
+- Produced when calculating the output type of a nested program
+  [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/interpreters/partial_eval.py#L2379)
+- Consumed when creating output tracers in the scope of an enclosing program
+  [source](https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/pjit.py#L1333)
+
+### Primitives and binding
+
+Primitives are like functions in Jaxpr. If the primitive has a body, then binding is associated with
+tracing of the program representing primitive’s body. The binding of such a primitive includes:
+  - Calculation of the input type based on the input arguments
+  - Creation of the inner tracing scope and the corresponding tracers.
+  - Calculation of the output type based on the results of the nested program.
+  - Creation of the resulting tracers in the caller's tracing scope.
+
+### Explicit/implicit arguments
+
+Separating explicit and implicit arguments makes sense when we trace Python program.  **Explicit**
+arguments/results are those which were explicitly mentioned in the source Python program. Implicit
+arguments are those that have to be added in order to fit into Jaxpr typing requirements.
+
+For example, in the following Python program:
+
+``` python
+def f(sz):
+  o = jnp.ones((sz+1,), dtype=float)
+  return o
+```
+
+We map the Python tracer `o` to the two Jaxpr variables `b:i64[], c:f[b]` to get the following
+equivalent Jaxpr program.
+
+```
+{ lambda ; a:i64[]. let
+    b:i64[] = add a 1
+    c:f64[b] = broadcast_in_dim[broadcast_dimensions=() shape=(None,)] 1.0 b
+  in (b, c) }
+```
+
+The first variable `b:i64[]` is the implicit result, while the second one is the explicit one.
```

Comment:
It's strange that JAX has no way of tracking this in the representation right? Or is it just a printer problem, and with the in-memory representation you can clearly distinguish between explicit and implicit arguments/results?

Reply:
The consequence is - several Python programs now might correspond to a single Jaxpr program. Yes, I think it might be a disadvantage, depending on their goals. They keep the implicitness information in types during the tracing, but they discard it at some point.